In [2]:
import pandas as pd
import numpy as np

In [54]:
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Dataset/Assignment/Project-CIP/census_income.csv')

In [27]:
df.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country annual_income  
0          2174             0              40  United-States         <=50K  
1             0             0              13  United-States         <=50K  
2             0             0              40  United-States         <=50K  
3             0             0              40  United-States         <=50K  
4             0             0              40           Cuba         <=50K

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  annual_income   32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


#Perform Basic Data cleaning Steps

In [55]:
df.isnull().sum().sum()

0

In [32]:
# This column data have ? and it could be a problem
df['workclass'].unique()

array(['State-gov', 'Self-emp-not-inc', 'Private', 'Federal-gov',
       'Local-gov', '?', 'Self-emp-inc', 'Without-pay', 'Never-worked'],
      dtype=object)

In [35]:
df['workclass'].value_counts()

workclass
Private             22696
Self-emp-not-inc     2541
Local-gov            2093
?                    1836
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: count, dtype: int64

In [36]:
# This column data looks good
df['education'].unique()

array(['Bachelors', 'HS-grad', '11th', 'Masters', '9th', 'Some-college',
       'Assoc-acdm', 'Assoc-voc', '7th-8th', 'Doctorate', 'Prof-school',
       '5th-6th', '10th', '1st-4th', 'Preschool', '12th'], dtype=object)

In [39]:
# This column data looks good
df['marital-status'].unique()

array(['Never-married', 'Married-civ-spouse', 'Divorced',
       'Married-spouse-absent', 'Separated', 'Married-AF-spouse',
       'Widowed'], dtype=object)

In [41]:
# This column data looks good
df['occupation'].unique()

array(['Adm-clerical', 'Exec-managerial', 'Handlers-cleaners',
       'Prof-specialty', 'Other-service', 'Sales', 'Craft-repair',
       'Transport-moving', 'Farming-fishing', 'Machine-op-inspct',
       'Tech-support', '?', 'Protective-serv', 'Armed-Forces',
       'Priv-house-serv'], dtype=object)

In [42]:
# This column data looks good
df['relationship'].unique()

array(['Not-in-family', 'Husband', 'Wife', 'Own-child', 'Unmarried',
       'Other-relative'], dtype=object)

In [43]:
# This column data looks good
df['race'].unique()

array(['White', 'Black', 'Asian-Pac-Islander', 'Amer-Indian-Eskimo',
       'Other'], dtype=object)

In [44]:
# This column data looks good
df['sex'].unique()

array(['Male', 'Female'], dtype=object)

In [45]:
# This column data have ? and it could be a problem
df['native-country'].unique()

array(['United-States', 'Cuba', 'Jamaica', 'India', '?', 'Mexico',
       'South', 'Puerto-Rico', 'Honduras', 'England', 'Canada', 'Germany',
       'Iran', 'Philippines', 'Italy', 'Poland', 'Columbia', 'Cambodia',
       'Thailand', 'Ecuador', 'Laos', 'Taiwan', 'Haiti', 'Portugal',
       'Dominican-Republic', 'El-Salvador', 'France', 'Guatemala',
       'China', 'Japan', 'Yugoslavia', 'Peru',
       'Outlying-US(Guam-USVI-etc)', 'Scotland', 'Trinadad&Tobago',
       'Greece', 'Nicaragua', 'Vietnam', 'Hong', 'Ireland', 'Hungary',
       'Holand-Netherlands'], dtype=object)

In [46]:
df['native-country'].value_counts()

native-country
United-States                 29170
Mexico                          643
?                               583
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
France                           29
Greece                           29
Ecuador                          28
Ireland                          24
Hong                             20
Cambodia                         19
Trinadad&Tobago                  19
Laos                             18
Thailand                         18
Yugoslavia                       16
Outlying-US(Guam-USVI-etc)       14
Honduras                         13
Hungary                          13
Scotland                         12
Holand-Netherlands                1
Name: count, dtype: int64

In [56]:
#We have found 2 columns having ? as value and we can't replace so we replacing it with null
df.replace('?', np.nan, inplace=True)

In [48]:
#Now we have null values in our data and we are going to drop it for cleaning data
df.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
annual_income        0
dtype: int64

In [57]:
# This we need for Question 3 in below section
df.isnull().sum().sum()

4262

In [58]:
print('Count before dropping null values',df.shape)
df.dropna(inplace=True)
print('Count before dropping null values',df.shape)


Count before dropping null values (32561, 15)
Count before dropping null values (30162, 15)


1. How many types of occupations do we have?
a. 13
b. 14
c. 15
d. 11

In [6]:
l=df['occupation'].unique()
print(len(l))
print(l)

15
['Adm-clerical' 'Exec-managerial' 'Handlers-cleaners' 'Prof-specialty'
 'Other-service' 'Sales' 'Craft-repair' 'Transport-moving'
 'Farming-fishing' 'Machine-op-inspct' 'Tech-support' '?'
 'Protective-serv' 'Armed-Forces' 'Priv-house-serv']


# 2: How many people are working as tech support and have an annual income greater than 50k?
# a. 278  (Answer)
# b. 389
# c. 289
# d. 934

In [52]:
df[(df['occupation']=='Tech-support') & (df['annual_income']=='>50K')]

age         workclass  fnlwgt     education  education-num  \
25      56         Local-gov  216851     Bachelors             13   
55      43           Private  237993  Some-college             10   
248     40           Private  207578    Assoc-acdm             12   
357     40           Private   96129     Assoc-voc             11   
469     30           Private  118551     Bachelors             13   
...    ...               ...     ...           ...            ...   
31510   42           Private  111483     Bachelors             13   
31653   25           Private  189656     Bachelors             13   
31807   51  Self-emp-not-inc  170086     Bachelors             13   
32076   59       Federal-gov  293971  Some-college             10   
32457   33           Private  139057       Masters             14   

           marital-status    occupation   relationship                race  \
25     Married-civ-spouse  Tech-support        Husband               White   
55     Married-civ-spouse  Tech-support        Husband               White   
248    Married-civ-spouse  Tech-support        Husband               White   
357    Married-civ-spouse  Tech-support        Husband               White   
469    Married-civ-spouse  Tech-support           Wife               White   
...                   ...           ...            ...                 ...   
31510  Married-civ-spouse  Tech-support        Husband               White   
31653       Never-married  Tech-support  Not-in-family               White   
31807  Married-civ-spouse  Tech-support        Husband               White   
32076  Married-civ-spouse  Tech-support        Husband               White   
32457  Married-civ-spouse  Tech-support        Husband  Asian-Pac-Islander   

          sex  capital-gain  capital-loss  hours-per-week native-country  \
25       Male             0             0              40  United-States   
55       Male             0             0              40  United-States   
248      Male             0          1977              60  United-States   
357      Male             0             0              40  United-States   
469    Female             0             0              16  United-States   
...       ...           ...           ...             ...            ...   
31510    Male             0             0              50  United-States   
31653    Male             0             0              60  United-States   
31807    Male             0             0              45  United-States   
32076    Male             0             0              40  United-States   
32457    Male             0             0              50  United-States   

      annual_income  
25             >50K  
55             >50K  
248            >50K  
357            >50K  
469            >50K  
...             ...  
31510          >50K  
31653          >50K  
31807          >50K  
32076          >50K  
32457          >50K  

[278 rows x 15 columns]

# 3: How many total missing values are present in the dataset?
# a. 4262  (Answer)
# b. 5000
# c. 5349
# d. 4302


In [53]:
#Here its coming zero because we have already dropped null in previous steps
total_missing = df.isnull().sum().sum()
print(total_missing)

0


# 4: If there are missing values in the Marital Status column, which option among the following should be used for replacing the missing values:
# a. Mean
# b. Median
# c. Mode   (Answer)
# d. All of the above

# Since 'Marital Status' is a categorical variable, the most appropriate option for replacing missing values is the Mode.


# 5: How many people are having private work classes and are not from the
# United States of America?
# a. 2151  (Answer)
# b. 2300
# c. 2000
# d. 2190


In [59]:
count = df[(df['workclass'] == 'Private') & (df['native-country'] != 'United-States')].shape[0]
print(count)


2151


# 6: How many people are either having Annual Income(last column) less than or
# equal to 50k or their working hours is greater than or equal to 40 hrs:
# a. 23008
# b. 23448
# c. 29505  (Answer)
# d. 25903




In [61]:
count = df[(df['annual_income'] <= '<=50K') | (df['hours-per-week'] >= 40)].shape[0]
print(count)

29505


# 7. Which of the following methods can you use for handling outliers?
# a. Interquartile Range(IQR) Method
# b. Z Score method
# c. Both of the above methods  (Answer)
# d. None of the above



In [ ]:
# Explain  IQR and Z-score methods

# IQR Method
# The IQR method identifies outliers based on the interquartile range.
# It calculates the difference between the 75th percentile (Q3) and the 25th percentile (Q1) of the data.
# Outliers are then identified as values that fall below Q1 - 1.5 * IQR or above Q3 + 1.5 * IQR.

# Example:
# Calculate the IQR for a specific column:
Q1 = df['age'].quantile(0.25)
Q3 = df['age'].quantile(0.75)
IQR = Q3 - Q1

# Identify outliers:
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
outliers_iqr = df[(df['age'] < lower_bound) | (df['age'] > upper_bound)]
print("Outliers using IQR method:\n", outliers_iqr)


# Z-score Method
# The Z-score method identifies outliers based on the standard deviation from the mean.
# It calculates the number of standard deviations a data point is away from the mean.
# Outliers are typically identified as values with a Z-score greater than 3 or less than -3.

# Example:
# Calculate the Z-score for a specific column:
mean = df['age'].mean()
std_dev = df['age'].std()
df['z_score'] = (df['age'] - mean) / std_dev

# Identify outliers:
outliers_zscore = df[(df['z_score'] > 3) | (df['z_score'] < -3)]
print("\nOutliers using Z-score method:\n", outliers_zscore)


# prompt: 8. Chi-square is used to analyze:
# a. Determine the relationship b/w the variables
# b. Compare observed results with expected results
# c. both a and b
# d. None of the above

# The answer is (c) both a and b

# Chi-square test is used to:
# a. Determine the relationship between the variables:
#    It helps to determine if there is a significant association between two categorical variables.
# b. Compare observed results with expected results:
#    It compares the observed frequencies in a contingency table with the expected frequencies under the assumption of independence.

In [ ]:
# Explain  Chi-square  with data example

import pandas as pd
from scipy.stats import chi2_contingency

# Sample data
data = {'Gender': ['Male', 'Male', 'Female', 'Female', 'Male', 'Female'],
        'Smoker': ['Yes', 'No', 'Yes', 'No', 'No', 'Yes']}
df = pd.DataFrame(data)

# Create a contingency table
contingency_table = pd.crosstab(df['Gender'], df['Smoker'])

# Perform the Chi-square test
chi2, p, dof, expected = chi2_contingency(contingency_table)

print("Chi-square statistic:", chi2)
print("P-value:", p)
print("Degrees of freedom:", dof)
print("Expected frequencies:\n", expected)

# Interpretation
alpha = 0.05  # Significance level

if p < alpha:
  print("There is a significant association between Gender and Smoker.")
else:
  print("There is no significant association between Gender and Smoker.")




# 9. What is VIF?
# a. It can detect multicollinearity
# b. If the VIF value is greater than 10, then there is no correlation between
# the independent variables
# c. It stands for Variance Impact Factor
# d. VIF is when there is no correlation between one predictor and the other
# predictors in a model.

# Answer: a. It can detect multicollinearity

# Explanation:
# VIF (Variance Inflation Factor) is a measure of multicollinearity in a regression model.
# It quantifies how much the variance of an estimated regression coefficient is increased because of multicollinearity.
# A high VIF value indicates that the predictor variable is highly correlated with other predictor variables in the model.
# Generally, a VIF value greater than 5 or 10 suggests a high level of multicollinearity.


# 10.What predict_proba will tell you?
# a. It will predict the class probabilities
# b. It will tell you the target value
# c. Both are correct
# d. None of the above

# The answer is (a) It will predict the class probabilities.

# predict_proba() in machine learning models gives you the probability of an observation belonging to each class.
# It's a helpful tool for understanding the model's confidence in its predictions.


# 11.Logistic regression is useful for regression problems:
# a. True
# b. False

# Logistic regression is a classification algorithm, not a regression algorithm. and Linear regression is used for regression problems.
# So the answer is (b) False.


# 12.In logistic regression, if the predicted logit is 0, what’s the
# transformed probability?
# a. 0.5
# b. 0.05
# c. Both of the above
# d. None of the above


In [ ]:

import math

def logit_to_probability(logit):
  """
  Transforms a logit value to a probability.
  """
  probability = 1 / (1 + math.exp(-logit))
  return probability

# Calculate the probability when the logit is 0
logit = 0
probability = logit_to_probability(logit)
print(probability)  # Output: 0.5

In [23]:
# prompt: 12.In logistic regression, if the predicted logit is 0, what’s the
# transformed probability?  what is the use of this with example with sample data

import pandas as pd
import math

def logit_to_probability(logit):
  """
  Transforms a logit value to a probability.
  """
  probability = 1 / (1 + math.exp(-logit))
  return probability

# Calculate the probability when the logit is 0
logit = 0
probability = logit_to_probability(logit)
print(probability)  # Output: 0.5

# Example with sample data
# Sample data
data = {'Feature1': [2, 3, 1, 4, 2, 5],
        'Feature2': [1, 2, 3, 1, 2, 3],
        'Target': [0, 1, 0, 1, 0, 1]}
df = pd.DataFrame(data)

# Assume you have a trained logistic regression model
# ... (Model training code here) ...

# Get the predicted logit for a new data point
new_data_point = [[2, 2]]
predicted_logit = model.predict_log_proba(new_data_point)

# Transform the logit to probability
probability = logit_to_probability(predicted_logit[0][1])  # Assuming the logit is for the positive class
print("Probability:", probability)

# Use case:
# In logistic regression, the output is a logit value, which represents the log odds of an event occurring.
# To get the actual probability, we need to transform the logit using the sigmoid function.
# This transformed probability helps us understand the likelihood of an event happening.
# For example, in a binary classification problem, a probability close to 1 indicates a high likelihood of belonging to the positive class,
# while a probability close to 0 indicates a high likelihood of belonging to the negative class.


0.5


# 13: Which variant of logistic regression is recommended when you have
# a categorical dependent variable with more than two values?
# a. Multiple Logistic regression
# b. Multinomial logistic regression
# c. Ordered logit regression
# d. Poisson regression

# The answer is (b) Multinomial logistic regression.

# Multinomial logistic regression is the appropriate variant of logistic regression when you have a categorical dependent variable with more than two values.
# It generalizes the binary logistic regression model to handle multiple categories.


#13 Perform the following tasks for answering the remaining questions

In [63]:
#Rename the last column as Annual Income
df = df.rename(columns={'annual_income':'Annual Income'})
#Remove the missing values from the dataset - This we have already performed

In [67]:
#Change the labels of categorical data into numerical data using Label Encoder.
#Split the dataset into a train and test of proportions 70:30 and set the random state to 0.
#Build a Logistic Regression Model on the data.

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import OneHotEncoder

# Select features and target variable
features = ['age', 'workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'hours-per-week', 'native-country']
target = 'Annual Income'

# Create a copy of the dataframe
df_model = df[features + [target]].copy()

# Handle missing values (replace with mode for categorical features)
for col in df_model.select_dtypes(include=['object']).columns:
  df_model[col] = df_model[col].fillna(df_model[col].mode()[0])

# Encode categorical features
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
encoded_features = encoder.fit_transform(df_model[features])
encoded_feature_names = encoder.get_feature_names_out(features)
encoded_df = pd.DataFrame(encoded_features, columns=encoded_feature_names)

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(encoded_df, df_model[target], test_size=0.3, random_state=42)

# Create and train the multinomial logistic regression model
model = LogisticRegression(multi_class='multinomial', solver='lbfgs', max_iter=1000)
model.fit(X_train, y_train)


LogisticRegression(max_iter=1000, multi_class='multinomial')

# 14.What is the accuracy score of the above model?
# a. 0.60 to 0.70
# b. 0.40 to 0.60
# c. 0.70 to 0.85  (Answer)
# d. None of the above

In [69]:
# Calculate the accuracy score
accuracy = model.score(X_test, y_test)
print(accuracy)


0.8393192617968837


# 15.What is the specificity of the above model?
# a. 0.20 to 0.30
# b. 0.30 to 0.40
# c. 0.50 to 0.60
# d. None of the above  (Answer)

In [70]:
from sklearn.metrics import confusion_matrix

# Predict the target variable for the test set
y_pred = model.predict(X_test)

# Calculate the confusion matrix
cm = confusion_matrix(y_test, y_pred)

# Calculate specificity
tn = cm[0, 0]
fp = cm[0, 1]
specificity = tn / (tn + fp)

print("Specificity:", specificity)


Specificity: 0.922713166839072


In [71]:
print(cm)

[[6244  523]
 [ 931 1351]]


# 16.What is the model’s precision when the target is False?
# a. 0.60 to 0.70
# b. 0.40 to 0.60
# c. 0.70 to 0.80
# d. None of the above

In [72]:
from sklearn.metrics import precision_score

# Calculate precision for the negative class (target is False)
precision_negative = precision_score(y_test, y_pred, pos_label='<=50K')

print("Precision for negative class:", precision_negative)


Precision for negative class: 0.8702439024390244


# 17.What is the total support value from the above model?
# a. 9049  (Asnwer) Support value is total number record provided for training
# b. 9032
# c. 10000
# d. 9847

# What is support value?

# Support value refers to the total number of actual instances for a specific class in the dataset.
# In the context of a confusion matrix, it represents the number of samples belonging to a particular class.

# For example, if you have a binary classification problem with classes "Positive" and "Negative",
# the support value for "Positive" would be the total number of actual "Positive" instances in the dataset.

# In the confusion matrix, the support values are usually displayed along the bottom row or rightmost column.
# They are important for understanding the distribution of classes in the dataset and evaluating the model's performance on each class.


In [73]:
import numpy as np
# Calculate the total support value
total_support = np.sum(cm)
print("Total support value:", total_support)


Total support value: 9049


# 18.What is the f1 score of the above model when the target is True?
# a. 0.30 to 0.40
# b. 0.40 to 0.50
# c. 0.60 to 0.70  (Answer)
# d. 0.90 to 0.99

In [74]:

from sklearn.metrics import f1_score

# Calculate the F1-score for the positive class (target is True)
f1_positive = f1_score(y_test, y_pred, pos_label='>50K')

print("F1-score for positive class:", f1_positive)


F1-score for positive class: 0.6501443695861405


# 19.How many records are correctly classified by the
# model?
# a. 7173
# b. 7043
# c. 7000
# d. None of the above  (Answer)

In [76]:
# Calculate the number of correctly classified records
correctly_classified = np.trace(cm)
print("Number of correctly classified records:", correctly_classified)
print(cm)


Number of correctly classified records: 7595
[[6244  523]
 [ 931 1351]]
